File: algorithms/python/lFUCache/lFUCache.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/lfu-cache/
# Author: Miao Zhang
# Date:   2021-02-13

class Node:
    def __init__(self, key, val, freq=0, pre=None, nxt=None):
        self.key = key
        self.val = val
        self.freq = freq
        self.pre = pre
        self.nxt = nxt
    
    def insert(self, nex):
        nex.pre = self
        nex.nxt = self.nxt
        self.nxt.pre = nex
        self.nxt = nex
    

def create_linked_list():
    head = Node(0, 0)
    tail = Node(0, 0)
    head.nxt = tail
    tail.pre = head
    return (head, tail)


class LFUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.min_freq = 0
        # freq---> list node
        self.freq_map = collections.defaultdict(create_linked_list)
        # key ---> node
        self.key_map = {}

    def get(self, key: int) -> int:
        if key not in self.key_map: return -1
        self.increase(self.key_map[key])
        return self.key_map[key].val

    def put(self, key: int, value: int) -> None:
        if self.capacity == 0: return
        if key in self.key_map:
            node = self.key_map[key]
            node.val = value
        else:
            node = Node(key, value)
            self.key_map[key] = node
            self.size += 1
        if self.size > self.capacity:
            self.size -= 1
            key_del = self.deleteNode(self.freq_map[self.min_freq][0].nxt)
            self.key_map.pop(key_del)
        self.increase(node)
        
    def deleteNode(self, node: Node):
        if node.pre:
            node.pre.nxt = node.nxt
            node.nxt.pre = node.pre
            if node.pre is self.freq_map[node.freq][0] and node.nxt is self.freq_map[node.freq][-1]:
                self.freq_map.pop(node.freq)
        return node.key
    
    def increase(self, node: Node):
        node.freq += 1
        self.deleteNode(node)
        self.freq_map[node.freq][-1].pre.insert(node)
        if node.freq == 1:
            self.min_freq = 1
        elif self.min_freq == node.freq - 1:
            head, tail = self.freq_map[node.freq - 1]
            if head.nxt is tail: self.min_freq = node.freq

        
# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: algorithms/cpp/lFUCache/lFUCache.cpp
// Source: https://leetcode.com/problems/lfu-cache/
// Author: Miao Zhang
// Date:   2021-02-13

struct CacheNode {
    int key;
    int value;
    int freq;
    // point to list location
    list<int>::const_iterator it;
};

class LFUCache {
public:
    LFUCache(int capacity): capacity_(capacity), min_freq_(0) {

    }

    int get(int key) {
        auto it = m_.find(key);
        if (it == m_.end()) return -1;
        touch(it->second);
        return it->second.value;
    }

    void put(int key, int value) {
        if (capacity_ == 0) return;

        auto it = m_.find(key);
        if (it != m_.end()) {
            it->second.value = value;
            touch(it->second);
            return;
        }

        if (m_.size() == capacity_) {
            int key_del = l_[min_freq_].back();
            l_[min_freq_].pop_back();
            m_.erase(key_del);
        }

        int freq = 1;
        min_freq_ = 1;
        l_[freq].push_front(key);
        m_[key] = {key, value, 1, l_[freq].cbegin()};
    }

private:
    int capacity_;
    int min_freq_;
    // key->CacheNode
    unordered_map<int, CacheNode> m_;
    // freq-> key
    unordered_map<int, list<int>> l_;

    void touch(CacheNode& node) {
        int prev_freq = node.freq;
        int new_freq = ++node.freq;
        l_[prev_freq].erase(node.it);
        if (l_[prev_freq].empty() && prev_freq == min_freq_) {
            l_.erase(prev_freq);
            min_freq_++;
        }
        l_[new_freq].push_front(node.key);
        node.it = l_[new_freq].cbegin();
    }
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache* obj = new LFUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
